File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int global_max = INT_MIN;
        dfs(root, global_max);
        return global_max;
    }
private:
    int dfs(TreeNode* node, int& global_max) {
        if (!node) {
            return 0;
        }

        int left_branch = std::max(0, dfs(node->left, global_max));
        int right_branch = std::max(0, dfs(node->right, global_max));
        global_max = std::max(global_max, node->val + left_branch + right_branch);
        return node->val + std::max(left_branch, right_branch);
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(node: Optional[TreeNode]) -> tuple(int,int):
            if not node:
                return (0,0)
            
            left_diameter, left_depth = dfs(node.left)
            right_diameter, right_depth = dfs(node.right)

            current_max_diameter = left_depth + right_depth
            global_max_diameter = max(left_diameter, right_diameter, current_max_diameter)
            current_depth = 1 + max(left_depth, right_depth)
            return (global_max_diameter, current_depth)
        final_diameter,_ = dfs(root)
        return final_diameter